flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/operations/MergeTableLikeUtilTest.java

```diff
@@ -431,6 +433,100 @@ public void mergeOverwritingMetadataColumnsDuplicate() {
         assertThat(mergedSchema, equalTo(expectedSchema));
     }
 
+    @Test
+    public void mergeIncludingMetadataColumnWithSameMetadataKey() {
+        TableSchema sourceSchema = TableSchema.builder().build();
+
+        List<SqlNode> derivedColumns =
+                Arrays.asList(
+                        metadataColumn("one", DataTypes.BOOLEAN(), true),
+                        metadataColumn("two", DataTypes.STRING(), "one", false));
+
+        Map<FeatureOption, MergingStrategy> mergingStrategies = getDefaultMergingStrategies();
+        mergingStrategies.put(FeatureOption.METADATA, MergingStrategy.INCLUDING);
+
+        thrown.expect(ValidationException.class);
+        thrown.expectMessage(
```

Comment:
How about we add this validation logic to `DefaultSchemaResolver`? This way we avoid checking duplicate metadata keys at three locations and fail as early as possible.